### PR TITLE
Add plugin framework-compatible location getter functions

### PR DESCRIPTION
**Pull request by SarahFrench**
# Background

This PR is proposing an idea for letting plugin framework code access location-type information for a resource.

## The old way of getting location information - using the SDK

When using the SDK it is possible to pass a [schema#ResourceData](https://pkg.go.dev/github.com/hashicorp/terraform/helper/schema#ResourceData) struct as an argument into a function. This struct is used to query and set the attributes of a resource, data source, or even the provider's configuration. This struct has methods like `Get`, `GetOk` and `Set` which take string arguments to identify the field in the resource's config. This means that you can write code where the field you are getting/setting can change depending on the value of the string being passed into those getter/setter functions. It also means that a function can handle a resource's configuration regardless of it's schema; there is a common interface between all resources and data sources based on the presence of `Get`/`GetOk`/`Set`/etc functions.

In the TPG/TPGB codebase there are functions that take arguments of type `schema.ResourceData` directly, or using an interface [TerraformResourceData](https://github.com/GoogleCloudPlatform/magic-modules/blob/e06e095cc480e8f82fc9ae76ebb05df7467150ca/mmv1/third_party/terraform/utils/utils.go#L30-L40), defined in the google package, that is coupled to the SDK:

```go
type TerraformResourceData interface {
	HasChange(string) bool
	GetOkExists(string) (interface{}, bool)
	GetOk(string) (interface{}, bool)
	Get(string) interface{}
	Set(string, interface{}) error
	SetId(string)
	Id() string
	GetProviderMeta(interface{}) error
	Timeout(key string) time.Duration
}
```

The util functions for getting location information for a resource relies on these features of the SDK package because it receives an argument of type `TerraformResourceData` (interface). See: [getRegion](https://github.com/hashicorp/terraform-provider-google/blob/0180090fba7e7132693aeb640bbf87d1900471c7/google/utils.go#L70), [getZone](https://github.com/hashicorp/terraform-provider-google/blob/15b2492b7192e4dcf03db4796080bf8a00239d3f/google/regional_utils.go#L36), [getLocation](https://github.com/hashicorp/terraform-provider-google/blob/15b2492b7192e4dcf03db4796080bf8a00239d3f/google/regional_utils.go#L21)

The way these functions are implemented makes them incompatible with plugin framework code, so they need to be re-written.

### Option 1. Make a function that can receive data describing resources from either SDK or plugin framework code:  Not possible

The first approach could be to identify a common interface between resource/data sources/etc in the SDK and the plugin framework to make a universally used version of the function. We would also need a common way to pass the provider's configuration into the function, as we use fallback values from the provider when considering location/region/zone.

This common interface doesn't exist; the SDK was able to have a common interface because it uses getter/setter patterns but the plugin framework encourages creation of custom structs to describe each resource/data source and accessing fields directly without getter/setter functions. Even within the plugin framework there is no common way to handle resources and datasources, and instead they must be defined with types from distinct packages: [github.com/hashicorp/terraform-plugin-framework/datasource](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.2.0/datasource) versus [github.com/hashicorp/terraform-plugin-framework/resource](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework@v1.2.0/resource).

In summary, this approach is not possible unless we write a lot of custom code to impose a common getter/setter pattern between SDK and plugin framework code.

### Option 2. Make a new function, to only use in plugin framework code, but make it receive the values directly as arguments - Possible but not ideal

Instead of getting a resource's region by using a function like:

```
getRegion(<argument containing all of resource's data>, <argument containing all of provider's configuration>)
```

...why not make a function like this that could be used regardless of how a resource is implemented:

```
getRegion(<value of resource's region field>, <value of default region in provider config>)
```

The problem here is that the logic in some functions is complex and requires access to more fields of a resource's schema than you might expect. In the case of the `getLocation` function it would need to be ready to receive 5 arguments: a resource's `location`, `region`, and `zone` value and the provider's default `region` and `zone` values. This approach results in function signatures that are very long and will always need to be passed in 'zero values' to indicate that a given field is not set (no resource has all 3 fields listed above).

```
getLocation("us-central1", "", "","us-central2", "us-central2-a") # poor readability
```

There is a risk that values could be passed in the incorrect order. There would also be a lot of effort when using this function to set up null values to pass in when a given field is not present in the resource's schema.

### Option 3. Use a new approach that isn't sticking closely to the old way this logic was implemented.

This PR attempts to solve this issue by making data model structs, which describe specific resources made using plugin framework, implement an interface that returns a struct that contains all location-related information related to that resource. Instead of creating new function like `getLocation` we create methods on that location struct that allow the information to be assessed and return any possible zone/region/location values.

Implementation of how that location object is constructed and returned is different depending on the resource. When code generation in Magic Modules is updated to use the plugin framework we will need to update templates so that the function is constructed differently depending on the resource:

```erb
# pseudo code templating code
func (m *<%= resource_name -%>Model) getLocationDescription(providerConfig *frameworkProvider) LocationDescription {
	return LocationDescription{
<%  if has_location -%>
	<%  if custom_location_schema_field -%>
		LocationSchemaField: types.StringValue("<%= custom_location_schema_field -%>"),
		ResourceLocation:    m.<%= custom_location_model_field -%>,
	<%  else -%>
		LocationSchemaField: types.StringValue("location"),
		ResourceLocation:    m.Location,
	<%  end -%>
<%  end -%>
<%  if has_region -%>
	<%  if custom_region_schema_field -%>
	        // do custom
	<%  else -%>
		RegionSchemaField: types.StringValue("region"),
		ResourceRegion:    m.Region,
	<%  end -%>
<%  end -%>
<%  if has_zone -%>
	<%  if custom_zone_schema_field -%>
	        // do custom
	<%  else -%>
		ZoneSchemaField:   types.StringValue("zone"),
		ResourceZone:      m.Zone,
         <%  end -%>
<%  end -%>
		ProviderRegion:    providerConfig.region,
		ProviderZone:      providerConfig.zone,
	}
}
```

templated code for a resource that only has 'region' information in its schema, and uses a custom name for that field: 

```erb
# pseudo templated code
func (m *FoobarModel) getLocationDescription(providerConfig *frameworkProvider) LocationDescription {
	return LocationDescription{
		RegionSchemaField: types.StringValue("custom_region_field"),
		ResourceRegion:    m.CustomRegionField,
		ProviderRegion:    providerConfig.region,
		ProviderZone:      providerConfig.zone,
	}
}
```

# In this PR

This PR makes an initial version of option 3 described above.

Tests are written for the methods on the `LocationDescription` struct to assert parity with the SDK code's `getLocation`, `getRegion` and `getZone` functions.

The PR also contains some refactoring of tests for the old `getLocation`, `getRegion` and `getZone` functions in response to reviewer feedback.

----

If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [x] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [x] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/run-provider-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:none

```

### CODE REVIEW

File: mmv1/third_party/terraform/data_sources/data_source_google_client_config.go
Comment (SarahFrench) [inline mmv1/third_party/terraform/data_sources/data_source_google_client_config.go:1]: This diff is to show how the new structs could be used in data sources/resource.

Note how `data.getLocationDescription` requires the provider config to be passed in, and this means that all data sources/resources would need to have the provider config passed in via their equivalent to:

```go
type GoogleClientConfigDataSource struct {
	providerConfig *frameworkProvider
}
```

File: mmv1/third_party/terraform/utils/framework_location.go [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+func (ld *LocationDescription) getRegion() (types.String, error) {
+	// Region from resource config
+	if !ld.ResourceRegion.IsNull() && !ld.ResourceRegion.IsUnknown() {
+		region := GetResourceNameFromSelfLink(ld.ResourceRegion.ValueString()) // Region could be a self link
+		return types.StringValue(region), nil
+	}
+	// Region from zone in resource config
+	if !ld.ResourceZone.IsNull() && !ld.ResourceZone.IsUnknown() {
+		region := getRegionFromZone(ld.ResourceZone.ValueString())
```
Comment (megan07) [inline mmv1/third_party/terraform/utils/framework_location.go:60]: do we need to worry about zone being a self link here?

File: mmv1/third_party/terraform/utils/framework_location_test.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+			},
+			ExpectedRegion: types.StringValue("us-central1"),
+		},
+		"returns a region derived from the zone field in resource config when region is unset": {
+			ld: LocationDescription{
+				ResourceZone: types.StringValue("provider-zone-a"),
+			},
+			ExpectedRegion: types.StringValue("provider-zone"),
+		},
+		"does not shorten region values when derived from a zone self link set in the resource config": {
```
Comment (megan07) [inline mmv1/third_party/terraform/utils/framework_location_test.go:93]: ah, this is the question i asked above. curious if this is intentional

File: mmv1/third_party/terraform/utils/framework_location_test.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+				ProviderZone: types.StringValue("provider-zone-a"),
+			},
+			ExpectedRegion: types.StringValue("provider-zone"),
+		},
+		"returns an error when zone values can't be found": {
+			ExpectedError: true,
+		},
+		"returns an error that mention non-standard schema field names when region value can't be found": {
+			ld: LocationDescription{
+				ZoneSchemaField: types.StringValue("foobar"),
```
Comment (megan07) [inline mmv1/third_party/terraform/utils/framework_location_test.go:116]: do we want this to be `RegionSchemaField`?

File: mmv1/third_party/terraform/utils/framework_location_test.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+				t.Fatalf("unexpected error using test: %s", err)
+			}
+			if region != tc.ExpectedRegion {
+				t.Fatalf("Incorrect region: got %s, want %s", region, tc.ExpectedRegion)
+			}
+		})
+	}
+}
+
+func TestLocationDescription_getLocation(t *testing.T) {
```
Comment (megan07) [inline mmv1/third_party/terraform/utils/framework_location_test.go:144]: Do we want any tests around if region and zone are set?

File: mmv1/third_party/terraform/utils/utils_test.go [hunk 5]
```diff
@@ -443,14 +468,35 @@ func TestGetRegion(t *testing.T) {
 			}
 
 			// Create resource config
-			// Here use ResourceComputeSubnetwork schema as example - because it has a region field in schema
+			// Here use a fictional schema as example because we need to have all of
+			// location, region, and zone fields present in the schema for the test,
+			// and no real resources would contain all of these
+			fictionalSchema := map[string]*schema.Schema{
```
Comment (megan07) [inline mmv1/third_party/terraform/utils/utils_test.go:474]: Can we pull this out of the different test runs and use this more globally?

**Review by megan07 — APPROVED** [review]
Looks great! Thanks!

File: mmv1/third_party/terraform/utils/framework_location_test.go (hunk 4 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location_test.go:154]: There are test cases covering when region and zone are used to get a location value, but I've updated the first test case to show what happens when _all_ values are available (even though that wouldn't/shouldn't really happen in the provider) - this is added in [f8dd164](https://github.com/GoogleCloudPlatform/magic-modules/pull/7835/commits/f8dd164310dcccb122b3609c7dcceece3a1d6078)

File: mmv1/third_party/terraform/utils/framework_location_test.go (hunk 3 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location_test.go]: Yes! Thanks for spotting this, addressed in 1c0c0c5

**Comment by SarahFrench** [comment]
rebased and force-pushed to address merge conflicts with main branch

File: mmv1/third_party/terraform/utils/framework_location.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+)
+
+type LocationDescriber interface {
+	getLocationDescription(providerConfig *frameworkProvider) LocationDescription
```
Comment (SarahFrench) [inline mmv1/third_party/terraform/utils/framework_location.go:11]: It could be that the function received the default region and zone values as arguments instead of the whole provider config.

```
getLocationDescription(providerRegion, providerZone types.String) LocationDescription {
	// do stuff
}
```

File: mmv1/third_party/terraform/utils/framework_location.go [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+		location := GetResourceNameFromSelfLink(ld.ResourceZone.ValueString()) // Zone could be a self link
+		return types.StringValue(location), nil
+	}
+
+	// Location from zone in provider config
+	if !ld.ProviderZone.IsNull() && !ld.ProviderZone.IsUnknown() {
+		return ld.ProviderZone, nil
+	}
+
+	return types.StringNull(), nil
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/framework_location.go:49]: should this return an error rather than nil?

File: mmv1/third_party/terraform/utils/framework_location.go (hunk 1 above)
Comment (melinath) [reply mmv1/third_party/terraform/utils/framework_location.go:60]: (FWIW it looks like this matches the current behavior: https://github.com/hashicorp/terraform-provider-google/blob/15b2492b7192e4dcf03db4796080bf8a00239d3f/google/field_helpers.go#L386 but I recognize that doesn't answer the question)

File: mmv1/third_party/terraform/utils/framework_location.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+}
+
+func (ld *LocationDescription) getRegion() (types.String, error) {
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/framework_location.go:52]: it looks like this doesn't replicate the GKE-specific behavior; I am guessing that the idea is that we would rewrite GKE in such a way that the workaround isn't necessary? (I like that idea, just want to confirm!)

File: mmv1/third_party/terraform/utils/framework_location_test.go [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+			ld: LocationDescription{
+				ResourceZone: types.StringValue("provider-zone-a"),
+			},
+			ExpectedRegion: types.StringValue("provider-zone"),
+		},
+		"does not shorten region values when derived from a zone self link set in the resource config": {
+			ld: LocationDescription{
+				ResourceZone: types.StringValue("https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1-a"),
+			},
+			ExpectedRegion: types.StringValue("https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1"), // Value isn't sortened from URI to name
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/framework_location_test.go:107]:
```suggestion
			ExpectedRegion: types.StringValue("https://www.googleapis.com/compute/v1/projects/my-project/zones/us-central1"), // Value isn't shortened from URI to name
```

File: mmv1/third_party/terraform/utils/framework_location_test.go (hunk 4 above)
Comment (melinath) [reply mmv1/third_party/terraform/utils/framework_location_test.go:154]: Could you also add a case where location is not available, but region and zone are, to make sure that region takes precedence if available? The current tests don't seem to cover that case.

File: mmv1/third_party/terraform/utils/framework_location_test.go [hunk 10]
```diff
@@ -0,0 +1,251 @@
...
+			},
+			ExpectedZone: types.StringValue("us-central1-a"),
+		},
+		"returns the value of the zone field in provider config when zone is unset in resource config": {
+			ld: LocationDescription{
+				ProviderZone: types.StringValue("provider-zone"),
+			},
+			ExpectedZone: types.StringValue("provider-zone"),
+		},
+		"returns an error when a zone value can't be found": {
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/framework_location_test.go:40]: could we add location / region values to this test (or as separate tests) to make sure they're ignored?

File: mmv1/third_party/terraform/utils/framework_location_test.go [hunk 11]
```diff
@@ -0,0 +1,251 @@
...
+			},
+			ExpectedRegion: types.StringValue("provider-region"),
+		},
+		"returns a region derived from the zone field in provider config when region unset in both resource and provider config": {
+			ld: LocationDescription{
+				ProviderZone: types.StringValue("provider-zone-a"),
+			},
+			ExpectedRegion: types.StringValue("provider-zone"),
+		},
+		"returns an error when zone values can't be found": {
```
Comment (melinath) [inline mmv1/third_party/terraform/utils/framework_location_test.go:121]: could we add a location value to this test (or as a separate test) to make sure it's ignored?

**Review by melinath — APPROVED** [review]
This implementation makes sense to me. :-) Overall LGTM - a few minor suggestions about tests to add & making more things exported.

File: mmv1/third_party/terraform/utils/framework_location.go [hunk 12]
```diff
@@ -0,0 +1,99 @@
+package google
```
[the remark this answers is not in the document]
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location.go:1]: Yeah I agree - I originally made this PR with a new location package as I imagined moving all the weirdness around locations stuff (e.g. util to get region from zone, util to get region from a regional self link) in a single package would help us make an inventory of all these utils. Currently they're spread between a few different files. But yeah I decide it was out of scope in the end.

I'll make the functions public and in future we can decide the correct lines to separate code by!

File: mmv1/third_party/terraform/utils/framework_location_test.go (hunk 10 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location_test.go:40]: Will do!

File: mmv1/third_party/terraform/utils/framework_location.go (hunk 1 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location.go:60]: This is addressed in a test case and there's currently parity with how the previous, SDK function handles a self link. 
I'd be interested in identifying ways to make the logic more consistent, as I agree that it's odd that zones as self links are accommodated in some places but not others, but also I don't think it's a good idea to make the function _more_ permissive than it already is

File: mmv1/third_party/terraform/utils/framework_location_test.go (hunk 2 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location_test.go:103]: Who knows, my hunch is that the zone-as-a-self-link use case was only relevant for a small number of APIs and only the util relevant to that resource(s) was updated

**Review by rileykarson — APPROVED** [review]
I haven't reviewed in depth (wanted to avoid too many cooks in the kitchen!) but the approach seems sound, LGTM

File: mmv1/third_party/terraform/utils/framework_location.go (hunk 1 above)
Comment (melinath) [reply mmv1/third_party/terraform/utils/framework_location.go:60]: +1 on standardizing this, if we can find a way to do it without breaking users (or maybe it could be a target for the next major release?)

File: mmv1/third_party/terraform/utils/framework_location.go (hunk 7 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location.go]: This was a big thing for me to miss 😓, thanks for spotting it! I've updated the function and updated the tests to fail if an error isn't returned when one is expected in [ef5f631](https://github.com/GoogleCloudPlatform/magic-modules/pull/7835/commits/ef5f6314b0cce05006ef92e24c4c00ce5271a960)

File: mmv1/third_party/terraform/data_sources/data_source_google_client_config.go [hunk 13]
```diff
@@ -34,6 +35,17 @@ type GoogleClientConfigModel struct {
 	AccessToken types.String `tfsdk:"access_token"`
 }
 
+func (m *GoogleClientConfigModel) getLocationDescription(providerConfig *frameworkProvider) LocationDescription {
+	return LocationDescription{
+		RegionSchemaField: types.StringValue("region"),
```
[the remark this answers is not in the document]
Comment (SarahFrench) [reply mmv1/third_party/terraform/data_sources/data_source_google_client_config.go:40]: That makes sense - addressed in [073b6aa](https://github.com/GoogleCloudPlatform/magic-modules/pull/7835/commits/073b6aa6538efcdeb4ef469f7438950a4108a944)

**Review by SarahFrench — COMMENTED** [review]
I managed to accidentally get my replies to review comments as my own review... submitting my accidental review to make them show

File: mmv1/third_party/terraform/utils/framework_location_test.go (hunk 11 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location_test.go]: Addressed in [0cbcbe8](https://github.com/GoogleCloudPlatform/magic-modules/pull/7835/commits/0cbcbe8cd547129dd4812c96c9ee97c14b6a617b)

Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location_test.go]: And also updated SDK tests in [757b007](https://github.com/GoogleCloudPlatform/magic-modules/pull/7835/commits/757b007f2f20c24e5ac9d04db624678666b54d03)

File: mmv1/third_party/terraform/utils/framework_location.go (hunk 1 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location.go:72]: Issue here: https://github.com/hashicorp/terraform-provider-google/issues/14503

File: mmv1/third_party/terraform/utils/framework_location.go (hunk 8 above)
Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location.go]: tldr; I think this is a problem to tackle closer to the time that these edge case resources are migrated.

---

From a quick search of the codebase where `getRegionFromSchema` is used (and `parseRegionalFieldValue`, which passes arguments into that function) I think that edge case is relevant in only a few places. I've almost certainly missed a few, but it's not very widespread from a first glace:

[google_redis_instance](https://github.com/hashicorp/terraform-provider-google/blob/14c4df39c94c4c1bc675241fbaf4376e4a5c8a8e/google/resource_redis_instance.go#L1966)

```go
region, err := getRegionFromSchema("region", "location_id", d, config)
```

[google_container_cluster](https://github.com/hashicorp/terraform-provider-google-beta/blob/4c61eb4c660d2379907224a3938b078b849a845c/google-beta/resource_container_cluster.go#L2056)

```go
subnetwork, err := parseRegionalFieldValue("subnetworks", v.(string), "project", "location", "location", d, config, true) // variant of ParseSubnetworkFieldValue
```

For `google_redis_instance` I think custom code is the answer. Either something new, or to implement the `getLocationDescription` function with `location_id` being used as the zone value (as currently a value of region is used over location_id).

For things that use `parseRegionalFieldValue` I think we need to tackle that when we move on to migrating that function. In the case above for `google_container_cluster` I think that it would need to have custom code that implements the `getLocationDescription` function and uses the value of the location field to set location, region and zone within the `LocationDescription` struct?

Comment (SarahFrench) [reply mmv1/third_party/terraform/utils/framework_location.go]: Please let me know if I'm overlooking something that I lack context for though! It's hard to foresee and unpick edge cases, especially if the current solution for those edge cases relies on features of the SDK that are missing in the plugin framework (accessing resource data via getters that can take string variables to identify fields with).

Comment (melinath) [reply mmv1/third_party/terraform/utils/framework_location.go]: Looking over the calling code - I think this code would only get triggered for google_container_cluster. Redis does have non-standard behavior but it still has different names for the two fields so wouldn't trigger this code block. I agree that the design you're proposing would let us handle this case in `GetLocationDescription` for the GKE resources. I had been worried that there might be non-obvious edge cases but it looks like everything is using hardcoded strings - so this seems like a fine way to proceed 👍  thanks for looking into it!

**Comment by SarahFrench** [comment]
The checks say there is an issue with generating terraform-google-conversion but I've been given the go-ahead to merge PRs despite that as it's not related to the content of this PR.
